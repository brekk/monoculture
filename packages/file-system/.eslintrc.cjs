module.exports = {
  root: true,
  extends: ['monoculture'],
  ignorePatterns: ['file-system.js'],
}
