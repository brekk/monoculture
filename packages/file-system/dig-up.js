import { Future } from 'fluture'
import { curry } from 'ramda'
import { findUp as __findUp } from 'find-up'

export const digUpWithCancel = curry(
  function _digUpWithCancel(cancel, opts, x) {
    return Future((bad, good) => {
      __findUp(x, opts)
        .catch(bad)
        .then(raw =>
          raw ? good(raw) : bad(new Error('No config file found!'))
        )
      return cancel
    })
  }
)
export const digUp = digUpWithCancel(() => {})
