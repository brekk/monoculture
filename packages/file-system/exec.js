import { pipe, curry, prop, ifElse, propOr, __ as $ } from 'ramda'
import { execa } from 'execa'
import { Future } from 'fluture'

export const didFail = propOr(true, 'failed')
export const fail = prop('stderr')
/* eslint-disable max-len */
/**
 * @pageSummary Call external processes and consume them as a Future-wrapped value.
 * This is a light wrapper around `execa`, so please see [execa](https://github.com/sindresorhus/execa)'s documentation for more info.
 */

/* eslint-disable jsdoc/tag-lines */
/**
 * Consume external commands as a Future-wrapped value.
 * @curried
 *  1. execWithConfig - Passes all possible configuration values plus a cancellation function.
 *
 *     @example
 *     ```js
 *     import { execWithConfig } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithConfig(
 *         function customCancellationFunction() {},
 *         'echo',
 *         { cleanup: true },
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *
 *  2. execWithCancel - Eschews any configuration and instead only expects a cancellation function, command and arguments.
 *
 *     @example
 *     ```js
 *     import { execWithCancel } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       execWithCancel(
 *         function customCancellationFunction() {},
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 *  3. exec - Eschews any configuration or cancellation function. Needs only command and arguments.
 *
 *     @example
 *     ```js
 *     import { exec } from 'file-system'
 *     import { fork } from 'fluture'
 *     fork(console.warn)(console.log)(
 *       exec(
 *         'echo',
 *         ['ahoy']
 *       )
 *     )
 *     ```
 */
/* eslint-enable max-len */
export const execWithConfig = curry(
  (cancellation, cmd, opts, args) =>
    new Future((bad, good) => {
      execa(cmd, args, opts)
        .catch(pipe(fail, bad))
        .then(ifElse(didFail, pipe(fail, bad), good))
      return cancellation
    })
)
export const execWithCancel = execWithConfig($, $, undefined)
export const exec = execWithCancel(() => {})
