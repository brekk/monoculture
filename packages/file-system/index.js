export * from './find-up'
export * from './exec'
export * from './fs'
export * from './interpret'
export * from './path'
