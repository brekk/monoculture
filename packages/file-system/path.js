import { join, normalize } from 'node:path'
import { curry } from 'ramda'

// pathJoin :: String -> String -> String
export const pathJoin = curry((pwd, x) => {
  // TODO: this should ostensibly be the province of a type system to ensure this safety
  if (typeof pwd !== 'string' || typeof x !== 'string') {
    throw new Error('Cannot normalize bad paths.')
  }
  return join(pwd, normalize(x))
})
