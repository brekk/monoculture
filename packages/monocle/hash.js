import crypto from 'node:crypto'

export const hash = buf => {
  const sum = crypto.createHash('sha256')
  sum.update(buf)
  return sum.digest('hex')
}
