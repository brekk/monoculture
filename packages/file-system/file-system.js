#!/usr/bin/env node

// src/flexeca.js
import { pipe, curry, ifElse, propOr } from "ramda";
import { execa } from "execa";
import { Future } from "fluture";
var didFail = propOr(true, "failed");
var fail = propOr("Something broke", "stderr");
var flexecaWithCanceller = curry(
  (cancellation, cmd, args) => new Future((bad, good) => {
    execa(cmd, args).catch(pipe(fail, (z) => `FAILED: "${z}"`, bad)).then(ifElse(didFail, pipe(fail, bad), good));
    return cancellation;
  })
);
var flexeca = flexecaWithCanceller(() => {
});

// src/fs.js
import {
  rm as __rm,
  mkdir as __mkdir,
  readFile as __readFile,
  writeFile as __writeFile,
  access as __access,
  constants
} from "node:fs";
import { propOr as propOr2, without, curry as curry2, pipe as pipe2, map } from "ramda";
import { chain, chainRej, Future as Future2, parallel } from "fluture";
import glob from "glob";
var localize = (z) => `./${z}`;
var readFile = (x) => new Future2((bad, good) => {
  __readFile(x, "utf8", (err, data) => err ? bad(err) : good(data));
  return () => {
  };
});
var readJSONFile = pipe2(readFile, map(JSON.parse));
var readDirWithConfig = curry2(
  (conf, g) => Future2((bad, good) => {
    try {
      glob(
        g,
        conf,
        (e, x) => (
          // thus far I cannot seem to ever call `bad` from within here
          e ? bad(e) : good(x)
        )
      );
    } catch (e) {
      bad(e);
    }
    return () => {
    };
  })
);
var readDir = readDirWithConfig({});
var writeFileWithConfig = curry2(
  (conf, file, content) => new Future2((bad, good) => {
    __writeFile(file, content, conf, (e) => {
      if (e) {
        bad(e);
        return;
      }
      good(content);
    });
    return () => {
    };
  })
);
var writeFile = writeFileWithConfig({ encoding: "utf8" });
var removeFileWithConfig = curry2(
  (options, fd) => new Future2((bad, good) => {
    __rm(fd, options, (err) => err ? bad(err) : good(fd));
    return () => {
    };
  })
);
var DEFAULT_REMOVAL_CONFIG = {
  force: false,
  maxRetries: 0,
  recursive: false,
  retryDelay: 100,
  parallel: 10
};
var removeFile = removeFileWithConfig(DEFAULT_REMOVAL_CONFIG);
var removeFilesWithConfig = curry2(
  (conf, list) => pipe2(
    map(removeFileWithConfig(without(["parallel"], conf))),
    parallel(propOr2(10, "parallel", conf))
  )(list)
);
var removeFiles = removeFilesWithConfig(DEFAULT_REMOVAL_CONFIG);
var mkdir = curry2(
  (conf, x) => new Future2((bad, good) => {
    __mkdir(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var mkdirp = mkdir({ recursive: true });
var access = curry2(
  (permissions, filePath) => new Future2((bad, good) => {
    __access(filePath, permissions, (err) => err ? bad(err) : good(true));
    return () => {
    };
  })
);
var exists = access(constants.F_OK);
var directoryOnly = (filePath) => filePath.slice(0, filePath.lastIndexOf("/"));
var writeFileWithAutoPath = curry2(
  (filePath, content) => pipe2(
    directoryOnly,
    (dir) => pipe2(
      exists,
      chainRej(() => mkdirp(dir))
    )(dir),
    chain(() => writeFile(filePath, content))
  )(filePath)
);
var rm = curry2(
  (conf, x) => new Future2((bad, good) => {
    __rm(x, conf, (err) => err ? bad(err) : good(x));
    return () => {
    };
  })
);
var rimraf = rm({ force: true, recursive: true });

// src/path.js
import { join, normalize } from "node:path";
import { curry as curry3 } from "ramda";
var pathRelativeTo = curry3((pwd, x) => {
  if (typeof pwd !== "string" || typeof x !== "string") {
    throw new Error("Cannot normalize bad paths.");
  }
  return join(pwd, normalize(x));
});

// src/interpret.js
import { Future as Future3 } from "fluture";
var interpret = (filepath) => Future3((bad, good) => {
  import(filepath).catch(bad).then(good);
  return () => {
  };
});
export {
  DEFAULT_REMOVAL_CONFIG,
  access,
  directoryOnly,
  exists,
  flexeca,
  flexecaWithCanceller,
  interpret,
  localize,
  mkdir,
  mkdirp,
  pathRelativeTo,
  readDir,
  readDirWithConfig,
  readFile,
  readJSONFile,
  removeFile,
  removeFileWithConfig,
  removeFiles,
  removeFilesWithConfig,
  rimraf,
  rm,
  writeFile,
  writeFileWithAutoPath,
  writeFileWithConfig
};
