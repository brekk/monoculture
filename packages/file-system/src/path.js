import { join, normalize } from 'node:path'
import { curry } from 'ramda'

// pathRelativeTo :: String -> String -> String
export const pathRelativeTo = curry((pwd, x) => join(pwd, normalize(x)))
