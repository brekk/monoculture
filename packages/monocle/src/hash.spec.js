import path from 'node:path'
import { fork } from 'fluture'
import { hash } from './hash'
import { readFile } from 'file-system'

test('hash', done => {
  fork(done)(raw => {
    // this is wildly self-documenting
    expect(hash(raw)).toMatchSnapshot()
    done()
  })(readFile(path.resolve(__dirname, 'hash.js')))
})
