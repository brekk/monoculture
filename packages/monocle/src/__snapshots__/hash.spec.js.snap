// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hash 1`] = `"5cd99d9df3d8e4c2fcc83cf248314a4836c71d88f8d6e4fdc0d84c4e8c7b029d"`;
